### DISCUSSION TITLE
How does dependabot know a nuget dependency is a "development" dependency?

### DISCUSSION THREAD

**Discussion by BinToss** [dependabot/dependabot-core#10285] [Q&A] [answered]
I've taken a quick look through the NugetUpdater helper projects, but I couldn't find an answer within 30 minutes.

The only information I could find was the old pull requests that added support for NuGet development dependencies when the implementation of NuGet ecosystem support was still Ruby-only. [And that implementation seems incorrect](https://github.com/dependabot/dependabot-core/pull/4671/files#diff-a9587b43fbeae71e602c2170f370aa3898b617ab25730cd53b381a3c7c5b7899). MSBuild/dotnet/et cetera does not have any implementation for "DevelopmentDependency" items as variants of PackageReference items.

This faulty support may have been implemented this way due to the [confusing and...poorly written documentation for NuGet's own DevelopmentDependency specification](https://github.com/NuGet/Home/wiki/DevelopmentDependency-support-for-PackageReference). I spent a few hours to cross-reference that spec doc to real world examples and what little other documentation is available to write [this](https://github.com/NuGet/Home/issues/4125#issuecomment-2249343141).

As stated in that write-up, setting the project configuration property "DevelopmentDependency" to `true` will do nothing but set the corresponding Nuspec property "developmentDependency" to `true`. When a `.nupkg`'s `.nuspec` has this property set to `true` and a package manager adds the package as a project dependency, the resulting dependency is written to the (SDK-style) project file as a PackageReference with two changes:
- The PackageReference's `PrivateAssets` metadata (xml attr or child element) is set to "all", the package dependency will not flow transitively to parent projects.
- The PackageReference's `ExcludeAssets` (default: "none") is set to "compile" if the nupkg has a `libs` directory containing .dll files. In some cases, this is written to set `IncludeAssets` (default: "all") to "runtime; build; native; contentfiles; analyzers". Those .dll files are libraries the project is built against and will require at runtime. They may be included as-is, trimmed down, and/or weaved into a single executable or library for distribution.

---

Search Keywords:
deps-dev nuget

**Comment by BinToss** [comment, marked as answer]
# 2024-07-28 | Analysis - Part 1

<details>

`CompatibilityChecker.CheckAsync` gets an `isDevDependency` bool from `CompatibilityChecker.GetPackageInfoAsync`.

https://github.com/dependabot/dependabot-core/blob/16f4c43f01b83aa6504c7fccd0a96ba3fda3cf5d/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Analyze/CompatabilityChecker.cs#L20-L33


In the body of `CompatibilityChecker.GetPackageInfoAsync`, `isDevDependency` is `true` if one of two conditions is met.
- The package's nuspec contains `<developmentDependency>true</developmentDependency>`. See https://github.com/NuGet/NuGet.Client/blob/5485ea697de98eee58746e0b0054cd478e33a1a5/src/NuGet.Core/NuGet.Packaging/Core/NuspecCoreReaderBase.cs#L123-L127. The package author may explicitly assign `false` to this property, but this is ignored if...
- The package has no TFMs (target frameworks) and no libs.
https://github.com/dependabot/dependabot-core/blob/16f4c43f01b83aa6504c7fccd0a96ba3fda3cf5d/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Analyze/CompatabilityChecker.cs#L73-L113

`CompatibilityChecker.CheckAsync` passes `isDevDependency` to `CompatibilityChecker.PerformCheck` wherein the method returns `true` if `isDevDependency && packageFrameworks.Length == 0`, among other things. `packageFrameworks` is an `ImmutableArray<NuGetFramework>` which are derived from the package the same way they are on NuGet.org.
https://github.com/dependabot/dependabot-core/blob/16f4c43f01b83aa6504c7fccd0a96ba3fda3cf5d/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Analyze/CompatabilityChecker.cs#L35-L49

Following up the call chain(s), you end up at `NuGetUpdater.Core.Analyze.AnalyzeWorker.RunAsync`

Examples: 
 - https://www.nuget.org/packages/GitVersion.MsBuild#supportedframeworks-body-tab. This package matches both conditions for `isDevDependency`. Its nuspec has `<developmentDependency>true</developmentDependency>` and it has no TFMs and no libs.

---

</details>

From what I can tell, whether or not a package is a development dependency is determined by a package dependency. Metadata which can be used to override this is not checked. A project may use non-development-dependency packages as development dependencies, but NuGetUpdater does not check for this. Primarily because there is no standard in the NuGet ecosystem for a dependent to explicitly state a dependency will only be used as a development dependency. It doesn't help that there are [multiple types of dependencies](https://github.com/dependabot/dependabot-core/blob/16f4c43f01b83aa6504c7fccd0a96ba3fda3cf5d/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/DependencyType.cs).

This is in stark contrast to the NPM ecosystem in which a project (i.e. [package.json](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#dependencies)) categorizes dependencies in multiple ways:
 - dependencies
 - devDependencies
 - peerDependencies (enhanced by separate peerDependenciesMeta)
 - bundleDependencies
 - optionalDependencies

~~Additionally, I still have not found any location at which a package's development dependency boolean is used by the commit message writer.~~

----

# 2024-07-31 | Analysis - Part 2



@brettfo 
> Your analysis in the previous comment is correct; we report something as a development dependency if we can't find any runtime files (e.g., anything under `lib/`, etc.)
> 
> As for surfacing that in the commit message, I'll have to yield to the core dependabot team, I'm not sure how that part of the code works.

I found it.

<details>

`Dependabot::PullRequestCreator::MessageBuilder.dependencies`
https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/common/lib/dependabot/pull_request_creator/message_builder.rb#L31-L32

`Dependabot::Dependency.production?`. If `false`, it's a dev dep. 
https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/common/lib/dependabot/dependency.rb#L180-L189
If top-level...
https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/common/lib/dependabot/dependency.rb#L191-L194
If NOT top-level...
https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/common/lib/dependabot/dependency.rb#L18-L26

<details><summary>Dependabot NuGet implementation</summary>

NuGet checks the dependency groups for a group whose `name` is "dependencies" and not "devDependencies".
https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/nuget/lib/dependabot/nuget.rb#L18-L26
See Line 117. A dependency is grouped in "devDependencies` if `NativeDependencyFileDiscovery` finds `dependency_details.is_dev_dependency` is true. Else, the dependency is grouped in "dependencies".
https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/nuget/lib/dependabot/nuget/native_discovery/native_dependency_file_discovery.rb#L104-L127

[NativeDependencyDetails](https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/nuget/lib/dependabot/nuget/native_discovery/native_dependency_details.rb) directly correlates to [NuGetUpdater.Core.Dependency](https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Dependency.cs)

</details>

`Dependabot::PullRequestCreator::MessageBuilder.pr_name_prefixer` calls `Dependabot::PullRequestCreator::PrNamePrefixer`.
https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/common/lib/dependabot/pull_request_creator/message_builder.rb#L809-L822

`Dependabot::PullRequestCreator::PrNamePrefixer.commit_prefix`
https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb#L91-L103

`Dependabot::PullRequestCreator::PrNamePrefixer.scope`
If nuget updates are configured with `commit-message: { include: scope }`, then this `scope` property determines the commit message's scope. It checks if any of the affected dependencies are `production?`. If `true`, the scope is `deps`. Else, `deps-dev`.
https://github.com/dependabot/dependabot-core/blob/8296fc29ba5ca1a7dbe628d56e457f2e438e3da3/common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb#L164-L166

</details>

**Comment by brettfo** [comment]
Your analysis in the previous comment is correct; we report something as a development dependency if we can't find any runtime files (e.g., anything under `lib/`, etc.)

As for surfacing that in the commit message, I'll have to yield to the core dependabot team, I'm not sure how that part of the code works.
